python/pyspark/tests.py

```diff
@@ -1175,6 +1175,32 @@ def test_pipe_functions(self):
         self.assertRaises(Py4JJavaError, rdd.pipe('grep 4', checkCode=True).collect)
         self.assertEqual([], rdd.pipe('grep 4').collect())
 
+    @unittest.skipIf(sys.version_info < (3, 2), "concurent.futures")
+    def test_async_actions(self):
+        data = [x for x in range(10)]
+        rdd = self.sc.parallelize(data)
+        f = rdd.collectAsync()
+        self.assertListEqual(f.result(), data)
+
+        f = rdd.countAsync()
+        self.assertEqual(f.result(), 10)
+
+        f = rdd.takeAsync(5)
+        self.assertEqual(f.result(), data[:5])
+
+        acc1 = self.sc.accumulator(0)
+
+        f = rdd.foreachAsync(lambda _: acc1.add(1))
+        self.assertTrue(
+            f.result() is None and acc1.value == len(data)
+        )
+
+        acc2 = self.sc.accumulator(0)
+        f = rdd.foreachPartitionAsync(lambda xs: [acc2.add(1) for _ in xs])
+        self.assertTrue(
+            f.result() is None and acc2.value == len(data)
```

Comment:
You should also do some tests with actual async execution, right now you block on each result before starting the next one.